core/primitives/src/stateless_validation/chunk_endorsement.rs

```diff
@@ -51,23 +65,78 @@ pub struct ChunkEndorsementV1 {
 }
 
 impl ChunkEndorsementV1 {
-    pub fn new(chunk_hash: ChunkHash, signer: &ValidatorSigner) -> ChunkEndorsementV1 {
+    pub fn new(chunk_hash: ChunkHash, signer: &ValidatorSigner) -> Self {
         let inner = ChunkEndorsementInner::new(chunk_hash);
         let account_id = signer.validator_id().clone();
         let signature = signer.sign_chunk_endorsement(&inner);
-        ChunkEndorsementV1 { inner, account_id, signature }
+        Self { inner, account_id, signature }
+    }
+
+    pub fn chunk_hash(&self) -> &ChunkHash {
+        &self.inner.chunk_hash
     }
 
     pub fn verify(&self, public_key: &PublicKey) -> bool {
         let data = borsh::to_vec(&self.inner).unwrap();
         self.signature.verify(&data, public_key)
     }
+}
 
-    pub fn chunk_hash(&self) -> &ChunkHash {
-        &self.inner.chunk_hash
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct ChunkEndorsementV2 {
+    // This is the part of the chunk endorsement that signed and included in the block header
+    inner: ChunkEndorsementInner,
+    // This is the signature of the inner field, to be included in the block header
+    signature: Signature,
+    // This consists of the metadata for chunk endorsement used in validation
+    metadata: ChunkEndorsementMetadata,
+    // Metadata signature is used to validate that the metadata is produced by the expected validator
+    metadata_signature: Signature,
+}
+
+impl ChunkEndorsementV2 {
+    pub fn new(
+        epoch_id: EpochId,
+        chunk_header: &ShardChunkHeader,
+        signer: &ValidatorSigner,
+    ) -> Self {
+        let inner = ChunkEndorsementInner::new(chunk_header.chunk_hash());
+        let metadata = ChunkEndorsementMetadata::V1(MetadataV1 {
+            account_id: signer.validator_id().clone(),
+            shard_id: chunk_header.shard_id(),
+            epoch_id,
+            height_created: chunk_header.height_created(),
+        });
+        let signature = signer.sign_chunk_endorsement(&inner);
+        let metadata_signature = signer.sign_chunk_endorsement_metadata(&metadata);
+        Self { inner, signature, metadata, metadata_signature }
+    }
+
+    // TODO(ChunkEndorsementV2): Remove this once we implement tracker_v2
+    pub fn into_v1(self) -> ChunkEndorsementV1 {
+        ChunkEndorsementV1 {
+            inner: self.inner,
+            account_id: match self.metadata {
+                ChunkEndorsementMetadata::V1(metadata) => metadata.account_id,
+            },
+            signature: self.signature,
+        }
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub enum ChunkEndorsementMetadata {
+    V1(MetadataV1),
```

Comment:
why do we want to have versioned metadata here? we already have version chunk endorsement, so we can just introduce V3 there if we need to make any changes, right?

Reply:
Yeah, I can remove the versioning for metadata.